assets/js/atomic/blocks/product-elements/button/frontend.tsx

```diff
@@ -0,0 +1,195 @@
+import { Cart } from '@woocommerce/type-defs/cart';
+import { createRoot } from '@wordpress/element';
+import NoticeBanner from '@woocommerce/base-components/notice-banner';
+type Context = {
+	woocommerce: {
+		isLoading: boolean;
+		addToCartText: string;
+		productId: number;
+		numberOfItems: number;
+		productPermalink: string;
+	};
+};
+
+type State = {
+	woocommerce: {
+		cart: Cart;
+		inTheCartText: string;
+	};
+};
+
+const storeNoticeClass = '.wc-block-store-notices';
+
+const createNoticeContainer = () => {
+	const noticeContainer = document.createElement( 'div' );
+	noticeContainer.classList.add( storeNoticeClass.replace( '.', '' ) );
+	return noticeContainer;
+};
+
+const injectNotice = ( domNode: Element, errorMessage: string ) => {
+	const root = createRoot( domNode );
+
+	root.render(
+		<NoticeBanner status="error" onRemove={ () => root.unmount() }>
+			{ errorMessage }
+		</NoticeBanner>
+	);
+
+	domNode?.scrollIntoView( {
+		behavior: 'smooth',
+		inline: 'nearest',
+	} );
+};
+
+const getProductById = ( cartState: Cart, productId: number ) => {
+	return cartState.items.find( ( item ) => item.id === productId );
+};
+
+const productButtonSelectors = {
+	woocommerce: {
+		addToCartText: ( {
+			context,
+			state,
+		}: {
+			context: Context;
+			state: State;
+		} ) => {
+			if ( context.woocommerce.numberOfItems === 0 ) {
+				return context.woocommerce.addToCartText;
+			}
+
+			if ( ! state.woocommerce.cart ) {
+				return state.woocommerce.inTheCartText.replace(
+					'###',
+					context.woocommerce.numberOfItems.toString()
+				);
+			}
+
+			const cartState = state.woocommerce.cart;
+			const product = getProductById(
+				cartState,
+				context.woocommerce.productId
+			);
+
+			if ( ! product ) {
+				return context.woocommerce.addToCartText;
+			}
+
+			return state.woocommerce.inTheCartText.replace(
+				'###',
+				product?.quantity?.toString()
+			);
+		},
+		isThereMoreThanOneItem: ( {
+			context,
+			state,
+		}: {
+			context: Context;
+			state: State;
+		} ) => {
+			if ( ! state.woocommerce?.cart ) {
+				return context.woocommerce.numberOfItems > 0;
+			}
+			const product = getProductById(
+				state.woocommerce.cart,
+				context.woocommerce.productId
+			);
+
+			return product !== undefined && product.quantity > 0;
+		},
+		isAdded: ( {
+			context,
+			selectors,
+			state,
+		}: {
+			context: Context;
+			// For now, let's just use any.
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+			selectors: any;
+			state: State;
+		} ) => {
+			return selectors.woocommerce.isThereMoreThanOneItem( {
+				context,
+				state,
+			} );
+		},
+	},
+};
+
+// @ts-expect-error: Store function isn't typed.
+store( {
+	selectors: productButtonSelectors,
+	actions: {
+		woocommerce: {
+			addToCart: async ( {
+				context,
+				ref,
+			}: {
+				context: Context;
+				state: State;
+				ref: HTMLElement;
+			} ) => {
+				if ( ! ref.classList.contains( 'ajax_add_to_cart' ) ) {
```

Comment:
I we do that, we would need to define a difference class name that we can use to know if the button needs to work with the interactivity API.

Reply:
https://github.com/woocommerce/woocommerce-blocks/pull/10006?notification_referrer_id=NT_kwDOAEQaRrI2ODkyNzI4NDY0OjQ0NjMxNzQ#discussion_r1258037507

As I said above, I think it is uncommon for old product grid blocks to be combined with a Products block. There are extensions/themes that could have specific behavior when `ajax_add_to_cart` is added. I think for compatibility reasons, it makes sense to keep the same class name.

Reply:
> There are extensions/themes that could have specific behavior when ajax_add_to_cart is added. I think for compatibility reasons, it makes sense to keep the same class name.

Oh, right, that makes sense.

> As I said above, I think it is uncommon for old product grid blocks to be combined with a Products block.

Ups, sorry if I missed this discussion. I have to say I'm not 100% sold on this part of the discussion. I think there are several use cases where a store owner might want to use the Products block alongside the Classic Template block, other WC product grid blocks or shortcodes. Ie: having the Product Catalog template with the Classic Template block but adding a banner on top with "Products on Sale" powered by the Products block. What do you think?

I don't want to block the whole PR because of this, but one thing we could explore is having a specific class for buttons using the Interactivity API (ie: `.wc-block-interactivity-api`) in addition to `.ajax_add_to_cart`. Then, updating this line in WC core:

https://github.com/woocommerce/woocommerce/blob/12e5edf03231abe6a3f98de002616a91cf369eca/plugins/woocommerce/client/legacy/js/frontend/add-to-cart.js#L63

To something like:

```JS
if ( $thisbutton.is( '.ajax_add_to_cart:not(.wc-block-interactivity-api)' ) ) {
```

(I didn't test it, so it might be more complex than what I think, but wanted to share this idea in case you think it makes sense)